### PR TITLE
HIVE-1638: Add cd.Spec.Platform.Azure.CloudName and plumb into deprovision.

### PR DESCRIPTION
**Pull request by abutcher**
https://issues.redhat.com/browse/HIVE-1638

/cc @joelddiaz

### CODE REVIEW

**Review by joelddiaz — COMMENTED** [review]
Since this is mainly to get AzureGov working perhaps this work can be delayed (assuming AzureGov also doesn't have a public DNS like AWSGov)... But, do we need to also update the DNSZone client stuff so that it sets appropriate alternative API endpoints when CloudName != "AzurePublicCloud"?

File: apis/hive/v1/azure/platform.go [hunk 1]
```diff
@@ -18,6 +18,39 @@ type Platform struct {
...
+	// GermanCloud is the Azure cloud environment used in Germany.
+	GermanCloud CloudEnvironment = "AzureGermanCloud"
+
+	// StackCloud is the Azure cloud environment used at the edge and on premises.
+	StackCloud CloudEnvironment = "AzureStackCloud"
+)
+
+// Name returns name that Azure uses for the cloud environment.
+// See https://github.com/Azure/go-autorest/blob/ec5f4903f77ed9927ac95b19ab8e44ada64c1356/autorest/azure/environments.go#L13
+func (e CloudEnvironment) Name() string {
```
[the remark this answers is not in the document]
Comment (abutcher) [reply apis/hive/v1/azure/platform.go:52]: I lifted this straight from the installer to make it easier to switch from `CloudEnvironment` to `string`.

File: apis/hive/v1/azure/platform.go [hunk 2]
```diff
@@ -18,6 +18,39 @@ type Platform struct {
...
+	USGovernmentCloud CloudEnvironment = "AzureUSGovernmentCloud"
+
+	// ChinaCloud is the Azure cloud environment used in China.
+	ChinaCloud CloudEnvironment = "AzureChinaCloud"
+
+	// GermanCloud is the Azure cloud environment used in Germany.
+	GermanCloud CloudEnvironment = "AzureGermanCloud"
+
+	// StackCloud is the Azure cloud environment used at the edge and on premises.
+	StackCloud CloudEnvironment = "AzureStackCloud"
```
[the remark this answers is not in the document]
Comment (abutcher) [reply apis/hive/v1/azure/platform.go:47]: We could limit this list to the supported Azure clouds.

File: apis/hive/v1/azure/platform.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (joelddiaz) [reply apis/hive/v1/azure/platform.go:52]: I guess it's fine. If we grow extra utility functions we can always revisit as this is not part of our public API.

File: apis/hive/v1/azure/platform.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (joelddiaz) [reply apis/hive/v1/azure/platform.go:47]: I guess to be on the safe side, lets drop AzureStackCloud. I'm sure the official request for AzureStack will come soon enough and we can address it more fully.

**Review by joelddiaz — COMMENTED** [review]
/lgtm

**Pull request merged**